content/blog/2022/2022-06-30-configuring-cors-with-spring.md

```diff
@@ -0,0 +1,430 @@
+ - Since maxAge = 3600, all pre-flight responses will be cached for 30 mins.
+
+#### Defining @CrossOrigin at method level
+
+````java
+    @CrossOrigin(origins = "http://localhost:4200", allowedHeaders = "Requestor-Type", exposedHeaders = "X-Get-Header")
+    @GetMapping
+    public ResponseEntity<List<BookDto>> getBooks(@RequestParam String type) {
+        HttpHeaders headers = new HttpHeaders();
+        headers.set("X-Get-Header", "ExampleHeader");
+        return ResponseEntity.ok().headers(headers).body(libraryService.getAllBooks(type));
+        }
+````
+
+ - Request coming from origin http://localhost:4200 only will be processed.
+ - If we expect only certain headers to be accepted, we can specify those headers in the allowedHeaders attribute. If the "Requestor-Type" header is not sent by the browser, the request will not be processed.
+ - If we set certain response headers, for the client application to be able to use them, we need to explicitly set the list of response headers to be exposed using the **exposedHeaders** attribute.
+
+#### Combination of @CrossOrigin at class and method levels
+
+````java
+    @CrossOrigin(maxAge = 3600)
+    @RestController
+    @RequestMapping("cors-library/managed/books")
+    public class LibraryController {
+
+        private static final Logger log = LoggerFactory.getLogger(LibraryController.class);
+
+        private final LibraryService libraryService;
+
+        public LibraryController(LibraryService libraryService) {
+            this.libraryService = libraryService;
+        }
+
+        @CrossOrigin(origins = "http://localhost:4200", allowedHeaders = "Requestor-Type")
+        @GetMapping
+        public ResponseEntity<List<BookDto>> getBooks(@RequestParam String type) {
+            HttpHeaders headers = new HttpHeaders();
+            headers.set("X-Get-Header", "ExampleHeader");
+            return ResponseEntity.ok().headers(headers).body(libraryService.getAllBooks(type));
+        }
+    }
+````
+By defining the annotation at both class and method levels
+    - its combined attributes will be applied to the methods i.e (origins, allowedHeaders, maxAge)
+
+### Enable CORS Configuration globally
+
+Instead of adding CORS to each of the resource separately, we could define a common CORS configuration that would apply to
+all resources defined. Here, we will use **WebMvcConfigurer** which is a part of the Spring Web MVC library.
+By overriding the **addCorsMapping** we will configure CORS to all URLs that are handled by Spring Web MVC.
+
+To define the same configuration (as explained in the previous sections) globally, we will use the configuration parameters
+defined in **application.yml** to create a bean as defined below:
+
+````yaml
+    web:
+    cors:
+      allowed-origins: "http://localhost:4200"
+      allowed-methods: GET, POST, PATCH, PUT, DELETE, OPTIONS, HEAD
+      max-age: 3600
+      allowed-headers: "Requestor-Type"
+      exposed-headers: "X-Get-Header"
```

Comment:
Are these custom properties you defined yourself and mapped to the `WebConfigProperties` object or are these pre-defined by Spring Boot? It would be good to explain that.